File: app-rn/src/core/trip.ts
// A trip is a pure state machine so the UI (SwiftUI or React) only renders it.
// Mirrors ios/TamilTrainCore/Sources/TamilTrainCore/Trip.swift.
import { dayKey } from './dates';
import { distractors, pickWord, shuffle } from './picker';
import { initialProgress, record } from './tracer';
import type { DayKey, Rng, Rules, TripSettings, Word, WordProgress } from './types';

export type StopPhase = 'waitingForGrownUp' | 'choosing' | 'solved';

export interface Stop {
  index: number;
  word: Word;
  kind: Word['kind'];
  options: Word[];          // empty for action stops
  phase: StopPhase;
  attempted: boolean;       // a wrong first tap has been recorded
  wrongIds: string[];
  outcome?: 'firstTry' | 'afterHint' | 'didAction' | 'triedAction';
}

export interface TripState {
  settings: TripSettings;
  today: DayKey;
  stopIndex: number;
  used: string[];
  stop: Stop | null;
  cargo: string[];          // word ids loaded onto the wagons
  stats: { pictureStops: number; firstTry: number; actionStops: number; didAction: number };
  finished: boolean;
}

export interface TripDeps {
  words: Word[];
  rules: Rules;
  rng: Rng;
  getProgress: () => Record<string, WordProgress>;
  saveProgress: (id: string, p: WordProgress) => void;
}

export function isActionStop(index: number, s: TripSettings, rules: Rules): boolean {
  return s.actionStops && index % rules.trip.actionStopEvery === rules.trip.actionStopEvery - 1;
}

export function startTrip(settings: TripSettings, deps: TripDeps, today: DayKey = dayKey()): TripState {
  const t: TripState = {
    settings, today, stopIndex: 0, used: [], stop: null, cargo: [],
    stats: { pictureStops: 0, firstTry: 0, actionStops: 0, didAction: 0 }, finished: false,
  };
  return makeStop(t, deps);
}

function makeStop(t: TripState, deps: TripDeps): TripState {
  if (t.stopIndex >= t.settings.length) return { ...t, stop: null, finished: true };
  const kind: Word['kind'] = isActionStop(t.stopIndex, t.settings, deps.rules) ? 'action' : 'picture';
  const word = pickWord({
    words: deps.words, progress: deps.getProgress(), used: new Set(t.used), stopIndex: t.stopIndex,
    today: t.today, scoring: deps.rules.scoring, picker: deps.rules.picker, rng: deps.rng,
  }, kind);
  if (!word) return { ...t, stop: null, finished: true };
  const options = kind === 'picture'
    ? shuffle([word, ...distractors(word, deps.words, t.settings.choices - 1, deps.rng)], deps.rng)
    : [];
  return {
    ...t,
    used: [...t.used, word.id],
    stop: { index: t.stopIndex, word, kind, options, phase: 'waitingForGrownUp', attempted: false, wrongIds: [] },
  };
}

/** Grown-up tapped "I said it". Options stay locked until then so he listens first. */
export function grownUpSaid(t: TripState): TripState {
  if (!t.stop || t.stop.phase !== 'waitingForGrownUp') return t;
  return { ...t, stop: { ...t.stop, phase: 'choosing' } };
}

function score(deps: TripDeps, word: Word, correct: boolean, guess: number, today: DayKey) {
  const cur = deps.getProgress()[word.id] ?? initialProgress(word.homeFrequency, deps.rules.scoring);
  deps.saveProgress(word.id, record(cur, correct, guess, today, deps.rules.scoring));
}

/** Child tapped a picture. Only the first tap at a stop is scored. */
export function tapPicture(t: TripState, id: string, deps: TripDeps): TripState {
  const s = t.stop;
  if (!s || s.kind !== 'picture' || s.phase !== 'choosing' || s.wrongIds.includes(id)) return t;
  const guess = 1 / s.options.length;
  if (id === s.word.id) {
    if (!s.attempted) score(deps, s.word, true, guess, t.today);
    return {
      ...t,
      cargo: [...t.cargo, s.word.id],
      stats: { ...t.stats, pictureStops: t.stats.pictureStops + (s.attempted ? 0 : 1), firstTry: t.stats.firstTry + (s.attempted ? 0 : 1) },
      stop: { ...s, phase: 'solved', outcome: s.attempted ? 'afterHint' : 'firstTry' },
    };
  }
  if (!s.attempted) score(deps, s.word, false, guess, t.today);
  return {
    ...t,
    stats: { ...t.stats, pictureStops: t.stats.pictureStops + (s.attempted ? 0 : 1) },
    stop: { ...s, attempted: true, wrongIds: [...s.wrongIds, id] },
  };
}

/** Grown-up reports whether he did the action without a demo. */
export function reportAction(t: TripState, did: boolean, deps: TripDeps): TripState {
  const s = t.stop;
  if (!s || s.kind !== 'action' || s.phase !== 'choosing') return t;
  score(deps, s.word, did, deps.rules.scoring.actionGuess, t.today);
  return {
    ...t,
    cargo: [...t.cargo, s.word.id],
    stats: { ...t.stats, actionStops: t.stats.actionStops + 1, didAction: t.stats.didAction + (did ? 1 : 0) },
    stop: { ...s, phase: 'solved', outcome: did ? 'didAction' : 'triedAction' },
  };
}

/** Swap the word at this stop (e.g. grown-up doesn't like it). Nothing is recorded. */
export function skipStop(t: TripState, deps: TripDeps): TripState {
  if (!t.stop || t.stop.phase === 'solved') return t;
  return makeStop({ ...t }, deps);
}

export function advance(t: TripState, deps: TripDeps): TripState {
  if (!t.stop || t.stop.phase !== 'solved') return t;
  return makeStop({ ...t, stopIndex: t.stopIndex + 1 }, deps);
}

File: ios/TamilTrainCore/Sources/TamilTrainCore/Trip.swift
import Foundation

// A trip is a plain value-type state machine; SwiftUI only renders it.
// Mirrors app-rn/src/core/trip.ts - keep the two in step.

public struct TripSettings: Codable, Equatable, Sendable {
    public var length: Int
    public var choices: Int
    public var actionStops: Bool
    public var showGloss: Bool

    public init(length: Int, choices: Int, actionStops: Bool, showGloss: Bool) {
        self.length = length
        self.choices = choices
        self.actionStops = actionStops
        self.showGloss = showGloss
    }
}

public struct Stop: Equatable, Sendable {
    public enum Phase: String, Equatable, Sendable { case waitingForGrownUp, choosing, solved }
    public enum Outcome: String, Equatable, Sendable { case firstTry, afterHint, didAction, triedAction }

    public let index: Int
    public let word: Word
    /// Empty for action stops.
    public let options: [Word]
    public var phase: Phase = .waitingForGrownUp
    /// A wrong first tap has been recorded.
    public var attempted = false
    public var wrongIds: [String] = []
    public var outcome: Outcome?

    public var kind: WordKind { word.kind }
}

public struct TripStats: Equatable, Sendable {
    public var pictureStops = 0
    public var firstTry = 0
    public var actionStops = 0
    public var didAction = 0
}

/// Side effects the trip needs, injected so tests can run without storage.
public struct TripDeps {
    public var words: [Word]
    public var rules: Rules
    public var rng: Rng
    public var getProgress: () -> [String: WordProgress]
    public var saveProgress: (String, WordProgress) -> Void

    public init(words: [Word], rules: Rules, rng: @escaping Rng,
                getProgress: @escaping () -> [String: WordProgress],
                saveProgress: @escaping (String, WordProgress) -> Void) {
        self.words = words
        self.rules = rules
        self.rng = rng
        self.getProgress = getProgress
        self.saveProgress = saveProgress
    }
}

public struct Trip: Equatable {
    public let settings: TripSettings
    public let today: DayKey
    public private(set) var stopIndex = 0
    public private(set) var used: [String] = []
    public private(set) var stop: Stop?
    /// Word ids loaded onto the wagons.
    public private(set) var cargo: [String] = []
    public private(set) var stats = TripStats()
    public private(set) var finished = false

    public init(settings: TripSettings, today: DayKey = Days.key(), deps: TripDeps) {
        self.settings = settings
        self.today = today
        makeStop(deps)
    }

    public static func isActionStop(_ index: Int, settings: TripSettings, rules: Rules) -> Bool {
        let every = rules.trip.actionStopEvery
        return settings.actionStops && every > 0 && index % every == every - 1
    }

    private mutating func makeStop(_ deps: TripDeps) {
        guard stopIndex < settings.length else {
            stop = nil
            finished = true
            return
        }
        let kind: WordKind = Trip.isActionStop(stopIndex, settings: settings, rules: deps.rules) ? .action : .picture
        let ctx = PickContext(words: deps.words, progress: deps.getProgress(), used: Set(used), stopIndex: stopIndex,
                              today: today, scoring: deps.rules.scoring, picker: deps.rules.picker, rng: deps.rng)
        guard let word = WordPicker.pickWord(ctx, kind: kind) else {
            stop = nil
            finished = true
            return
        }
        let options = kind == .picture
            ? WordPicker.shuffle([word] + WordPicker.distractors(for: word, from: deps.words, count: settings.choices - 1, rng: deps.rng), rng: deps.rng)
            : []
        used.append(word.id)
        stop = Stop(index: stopIndex, word: word, options: options)
    }

    private func score(_ word: Word, correct: Bool, guess: Double, deps: TripDeps) {
        let current = deps.getProgress()[word.id] ?? Tracer.initial(homeFrequency: word.homeFrequency, rules: deps.rules.scoring)
        deps.saveProgress(word.id, Tracer.record(current, correct: correct, guess: guess, day: today, rules: deps.rules.scoring))
    }

    /// Grown-up tapped "I said it". Pictures stay locked until then so he listens first.
    public mutating func grownUpSaid() {
        guard stop?.phase == .waitingForGrownUp else { return }
        stop?.phase = .choosing
    }

    /// Child tapped a picture. Only the first tap at a stop is scored.
    public mutating func tapPicture(_ id: String, deps: TripDeps) {
        guard var s = stop, s.kind == .picture, s.phase == .choosing, !s.wrongIds.contains(id) else { return }
        let guess = 1.0 / Double(s.options.count)
        if id == s.word.id {
            if !s.attempted {
                score(s.word, correct: true, guess: guess, deps: deps)
                stats.pictureStops += 1
                stats.firstTry += 1
            }
            cargo.append(s.word.id)
            s.phase = .solved
            s.outcome = s.attempted ? .afterHint : .firstTry
        } else {
            if !s.attempted {
                score(s.word, correct: false, guess: guess, deps: deps)
                stats.pictureStops += 1
            }
            s.attempted = true
            s.wrongIds.append(id)
        }
        stop = s
    }

    /// Grown-up reports whether he did the action without a demo.
    public mutating func reportAction(did: Bool, deps: TripDeps) {
        guard var s = stop, s.kind == .action, s.phase == .choosing else { return }
        score(s.word, correct: did, guess: deps.rules.scoring.actionGuess, deps: deps)
        stats.actionStops += 1
        if did { stats.didAction += 1 }
        cargo.append(s.word.id)
        s.phase = .solved
        s.outcome = did ? .didAction : .triedAction
        stop = s
    }

    /// Swap the word at this stop. Nothing is recorded.
    public mutating func skip(deps: TripDeps) {
        guard let s = stop, s.phase != .solved else { return }
        makeStop(deps)
    }

    public mutating func advance(deps: TripDeps) {
        guard stop?.phase == .solved else { return }
        stopIndex += 1
        makeStop(deps)
    }
}
